src/Text/Pandoc/App/FormatHeuristics.hs

```diff
@@ -15,16 +15,13 @@ module Text.Pandoc.App.FormatHeuristics
   ) where
 
 import Data.Char (toLower)
+import Data.Semigroup (First(First), getFirst)
 import Data.Text (Text)
 import System.FilePath (takeExtension)
 
 -- Determine default format based on file extensions.
 formatFromFilePaths :: [FilePath] -> Maybe Text
-formatFromFilePaths [] = Nothing
-formatFromFilePaths (x:xs) =
-  case formatFromFilePath x of
-    Just f     -> Just f
-    Nothing    -> formatFromFilePaths xs
+formatFromFilePaths = fmap getFirst . foldMap (fmap First . formatFromFilePath)
 
```

Comment:
I'm curious, what would you say is the advantage of this change?  Although it shortens the code, it's now harder to see what it does.  It's also less obvious that it stops processing on the first match (though you could figure this out if you were good at reasoning about laziness in Haskell).